### ISSUE TITLE
Show auto-completion for user-defined methods when overriding in GDScript

### ISSUE THREAD

**Issue by poohcom1** [godotengine/godot-proposals#5896]
### Describe the project you are working on

A 2D action game with a lot of interface-like classes

### Describe the problem or limitation you are having in your project

Overriding user-defined methods are cumbersome as there are no auto-completion.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

In the script editor, show auto-completion for user-defined methods when declaring/overriding functions with the `func` keyword.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In parent class:
![image](https://user-images.githubusercontent.com/74857873/205494011-d7d33206-5f96-4082-89d3-47f59aa89c91.png)

In child class:
![image](https://user-images.githubusercontent.com/74857873/205494000-09ff942e-b98f-4bae-bf1b-a1491b892054.png)

Specifics:
 - Suggestions will include all methods in all user-defined super classes. Unless we have virtual methods (#1631) or some form of access modifier, all methods are technically override-able, therefore it is logical to autocomplete all methods.
 - On selection, inserted method should behave like it does currently with virtual methods, i.e. include parameters (and type-hints if `text_editor/completion/add_type_hints` is on).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This involves core GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

This involves core GDScript.

**Comment by poohcom1** [comment]
I have a [working implementation](https://github.com/poohcom1/godot/commit/4270db15b6380146351e39bd4f4b0d7d4d9d324d) of this, but it's blocked by a few issues:

 - https://github.com/godotengine/godot/issues/69050
   - Auto-completed custom classes in parent classes are broken
 - https://github.com/godotengine/godot/pull/68086
   - Refactor that blocks the above issue and could affect this proposal's implementation

**Comment by KoBeWi** [comment]
Technically a duplicate of #393, but you have better mock-ups and even have an implementation, so I'll close my proposal.

IMO we shouldn't show all methods for override, but there should be a way to explicitly mark them as virtual. Either by `_` convention or via some new annotation.

**Comment by poohcom1** [comment]
> IMO we shouldn't show all methods for override, but there should be a way to explicitly mark them as virtual. Either by `_` convention or via some new annotation.

Personally, I think keywords/annotations can wait until we have actual virtual functions. Given how previous discussions weren't really going anywhere (and people couldn't really agree on the keyword), I wanted this proposal to just be a small QoL improvement that can be easily added to Godot. Because of that, showing all methods seem to be the most 'natural' behavior.

As for `_`, I'm kinda on the fence. I feel like it isn't a good idea to lock features behind naming conventions unless it has an actual effect on the interpreter (like the Python double underscore). If the concern is that suggestions will get filled with unnecessary methods, I think users can come up with their own convention, and rely on the filtering to get the method they need. That said, I'm still okay with `_` since it does follow the existing conventions and wouldn't be so hard to discover. Let me know what you think.

**Comment by KoBeWi** [comment]
Actually, maybe showing everything wouldn't be that bad. Built-in virtual methods always start with `_`, so we could just sort the list in a way that makes them always listed on top. Users could also take advantage of this and use the `_` convention to have their "virtual" methods easily available.

To me the most important part is that autocompleting a method automatically adds all necessary arguments. I use overriding a lot and it's a pain, because I always have to lookup the declarations and copy them. I wouldn't really mind the method list suddenly getting long.

btw if you have implementation ready, you can open a PR even if it depends on something else.

**Comment by poohcom1** [comment]
@KoBeWi I've been cleaning up the implementation a bit so I can raise a PR, but one main issue I'm stuck on is with default parameters. Since GDScript is allows any expression as default parameters, there are some edge cases where you can't really reduce the expression (like function calls). I've been using a modified `gdscript_editor::_make_arguments_hint` to generate the arguments and types, but it prints `<unknown>` for default parameters if the value couldn't be reduced.

**For Example:**
In parent:
![image](https://user-images.githubusercontent.com/74857873/205561346-359de4e4-35c7-4a50-b24c-bd201eacbf8c.png)
In child:
![image](https://user-images.githubusercontent.com/74857873/205561323-845a4514-6de3-4e33-ab0e-64fd06456263.png)

This also applies for enums and arrays/dictionaries (possibly a bug).

Because of this, I thought it might just be better to just copy over the exact default value expression? Given that child classes should share all members of their parents, expressions that are correct in parents should in theory also be correct in child classes.

Any thought?

**Comment by OrganicPepper** [comment]
This would be super lovely to have, also perhaps options in GDScript project settings to warn / error when overriding a parent function incorrectly.

**Comment by KoBeWi** [comment]
> warn / error when overriding a parent function incorrectly.

This already exists.

**Comment by OrganicPepper** [comment]
Ah, so it does, my mistake! It wasn't working earlier, but I had a compilation error elsewhere which was preventing the overriding error from showing!
